File: app/src/main/java/org/docheinstein/minimote/packet/MinimotePacketFactory.java
package org.docheinstein.minimote.packet;

import org.docheinstein.minimote.keys.MinimoteKeyType;
import org.docheinstein.minimote.utils.ByteUtils;

import java.util.List;

public class MinimotePacketFactory {
    /*
        LeftDown(0x01),
        LeftUp(0x02),
        LeftClick(0x03),
        MiddleDown(0x04),
        MiddleUp(0x05),
        MiddleClick(0x06),
        RightDown(0x07),
        RightUp(0x08),
        RightClick(0x09),
        ScrollDown(0x0A),
        ScrollUp(0x0B),
        Move(0x0C),
        Write(0x0D),
        KeyDown(0x0E),
        KeyUp(0x0F),
        KeyClick(0x10),

        DiscoverRequest(0xFE),
     */

    public static MinimotePacket newLeftDown() {
        return new MinimotePacket(MinimotePacketType.LeftDown);
    }

    public static MinimotePacket newLeftUp() {
        return new MinimotePacket(MinimotePacketType.LeftUp);
    }

    public static MinimotePacket newLeftClick() {
        return new MinimotePacket(MinimotePacketType.LeftClick);
    }

    public static MinimotePacket newMiddleDown() {
        return new MinimotePacket(MinimotePacketType.MiddleDown);
    }

    public static MinimotePacket newMiddleUp() {
        return new MinimotePacket(MinimotePacketType.MiddleUp);
    }

    public static MinimotePacket newMiddleClick() {
        return new MinimotePacket(MinimotePacketType.MiddleClick);
    }

    public static MinimotePacket newRightDown() {
        return new MinimotePacket(MinimotePacketType.RightDown);
    }

    public static MinimotePacket newRightUp() {
        return new MinimotePacket(MinimotePacketType.RightUp);
    }

    public static MinimotePacket newRightClick() {
        return new MinimotePacket(MinimotePacketType.RightClick);
    }

    public static MinimotePacket newScrollDown() {
        return new MinimotePacket(MinimotePacketType.ScrollDown);
    }

    public static MinimotePacket newScrollUp() {
        return new MinimotePacket(MinimotePacketType.ScrollUp);
    }

    public static MinimotePacket newMove(int mid, int x, int y) {
        final int PAYLOAD_SIZE = 4;

        // | MID (8 bit) | X (12 bit) | Y (12 bit) |
        int payloadValue =
                ((mid & 0xFF) << 24) |
                ((x & 0xFFF) << 12) |
                ((y & 0xFFF));

        byte[] payload = new byte[PAYLOAD_SIZE];
        ByteUtils.put32(payloadValue, payload);

        return new MinimotePacket(
                MinimotePacket.HEADER_SIZE + PAYLOAD_SIZE,
                MinimotePacketType.Move,
                payload
        );
    }

    public static MinimotePacket newWrite(char ch) {
        final int PAYLOAD_SIZE = 1;

        // | CHAR (8 bit) |
        int payloadValue = (int) ch;

        byte[] payload = new byte[PAYLOAD_SIZE];
        ByteUtils.put8(payloadValue, payload);

        return new MinimotePacket(
                MinimotePacket.HEADER_SIZE + PAYLOAD_SIZE,
                MinimotePacketType.Write,
                payload
        );
    }

    public static MinimotePacket newKeyDown(MinimoteKeyType keyType) {
        final int PAYLOAD_SIZE = 1;

        // | KEY TYPE (8 bit) |
        int payloadValue = keyType.getValue();

        byte[] payload = new byte[PAYLOAD_SIZE];
        ByteUtils.put8(payloadValue, payload);

        return new MinimotePacket(
                MinimotePacket.HEADER_SIZE + PAYLOAD_SIZE,
                MinimotePacketType.KeyDown,
                payload
        );
    }

    public static MinimotePacket newKeyUp(MinimoteKeyType keyType) {
        final int PAYLOAD_SIZE = 1;

        // | KEY TYPE (8 bit) |
        int payloadValue = keyType.getValue();

        byte[] payload = new byte[PAYLOAD_SIZE];
        ByteUtils.put8(payloadValue, payload);

        return new MinimotePacket(
                MinimotePacket.HEADER_SIZE + PAYLOAD_SIZE,
                MinimotePacketType.KeyUp,
                payload
        );
    }

    public static MinimotePacket newKeyClick(MinimoteKeyType keyType) {
        final int PAYLOAD_SIZE = 1;

        // | KEY TYPE (8 bit) |
        int payloadValue = keyType.getValue();

        byte[] payload = new byte[PAYLOAD_SIZE];
        ByteUtils.put8(payloadValue, payload);

        return new MinimotePacket(
                MinimotePacket.HEADER_SIZE + PAYLOAD_SIZE,
                MinimotePacketType.KeyClick,
                payload
        );
    }

    public static MinimotePacket newHotkey(List<MinimoteKeyType> keys) {
        final int PAYLOAD_SIZE = keys.size();

        // | KEY TYPE (8 bit) * keys.length |
        byte[] payload = new byte[PAYLOAD_SIZE];

        int i = 0;
        for (MinimoteKeyType key : keys) {
            ByteUtils.put8(key.getValue(), payload, i);
            i++;
        }

        return new MinimotePacket(
                MinimotePacket.HEADER_SIZE + PAYLOAD_SIZE,
                MinimotePacketType.Hotkey,
                payload
        );
    }

    public static MinimotePacket newDiscoverRequest() {
        return new MinimotePacket(MinimotePacketType.DiscoverRequest);
    }

}

File: app/src/main/java/org/docheinstein/minimotek/packet/MinimotePacketFactory.kt
package org.docheinstein.minimotek.packet


import org.docheinstein.minimotek.keys.MinimoteKeyType
import org.docheinstein.minimotek.util.set16
import org.docheinstein.minimotek.util.set32
import org.docheinstein.minimotek.util.set8

/**
 * Factory of [MinimotePacket]s for each possible [MinimotePacketType].
 */
object MinimotePacketFactory {
    fun newLeftDown(): MinimotePacket {
        return MinimotePacket(MinimotePacketType.LeftDown)
    }

    fun newLeftUp(): MinimotePacket {
        return MinimotePacket(MinimotePacketType.LeftUp)
    }

    fun newLeftClick(): MinimotePacket {
        return MinimotePacket(MinimotePacketType.LeftClick)
    }

    fun newMiddleDown(): MinimotePacket {
        return MinimotePacket(MinimotePacketType.MiddleDown)
    }

    fun newMiddleUp(): MinimotePacket {
        return MinimotePacket(MinimotePacketType.MiddleUp)
    }

    fun newMiddleClick(): MinimotePacket {
        return MinimotePacket(MinimotePacketType.MiddleClick)
    }

    fun newRightDown(): MinimotePacket {
        return MinimotePacket(MinimotePacketType.RightDown)
    }

    fun newRightUp(): MinimotePacket {
        return MinimotePacket(MinimotePacketType.RightUp)
    }

    fun newRightClick(): MinimotePacket {
        return MinimotePacket(MinimotePacketType.RightClick)
    }

    fun newScrollDown(): MinimotePacket {
        return MinimotePacket(MinimotePacketType.ScrollDown)
    }

    fun newScrollUp(): MinimotePacket {
        return MinimotePacket(MinimotePacketType.ScrollUp)
    }

    fun newMove(mid: Int, x: Int, y: Int): MinimotePacket {
        // | MOVE_ID (8 bit) | X (12 bit) | Y (12 bit) |
        val payload = ByteArray(4)
        payload.set32(
            ((mid and 0xFF) shl 24) or
            ((x and 0xFFF) shl 12) or
            ((y and 0xFFF))
        )
        return MinimotePacket(MinimotePacketType.Move, payload)
    }

    fun newWrite(c: Char): MinimotePacket {
        // | UNICODE (4 byte) |
        val payload = ByteArray(4)
        payload.set32(c.code)
        return MinimotePacket(MinimotePacketType.Write, payload)
    }


    fun newKeyDown(keyType: MinimoteKeyType): MinimotePacket {
        // | KEY TYPE (8 bit) |
        val payload = ByteArray(1)
        payload.set8(keyType.value)
        return MinimotePacket(MinimotePacketType.KeyDown, payload)
    }

    fun newKeyUp(keyType: MinimoteKeyType): MinimotePacket {
        // | KEY TYPE (8 bit) |
        val payload = ByteArray(1)
        payload.set8(keyType.value)
        return MinimotePacket(MinimotePacketType.KeyUp, payload)
    }

    fun newKeyClick(keyType: MinimoteKeyType): MinimotePacket {
        // | KEY TYPE (8 bit) |
        val payload = ByteArray(1)
        payload.set8(keyType.value)
        return MinimotePacket(MinimotePacketType.KeyClick, payload)
    }

    fun newHotkey(keys: List<MinimoteKeyType>): MinimotePacket {
        // | KEY TYPE (8 bit) * keys.size |
        val payload = ByteArray(keys.size)
        var i = 0
        for (key in keys) {
            payload.set8(key.value, offset = i)
            i++
        }
        return MinimotePacket(MinimotePacketType.Hotkey, payload)
    }

    fun newDiscoverRequest(): MinimotePacket {
        return MinimotePacket(MinimotePacketType.DiscoverRequest)
    }

    fun newPing(recvPort: Int): MinimotePacket {
        // | RECV_PORT (16 bit) |
        val payload = ByteArray(2)
        payload.set16(recvPort)
        return MinimotePacket(MinimotePacketType.Ping, payload)
    }
}